src/nvim/generators/gen_events.lua

```diff
@@ -25,25 +25,22 @@ static const struct event_name {
 } event_names[] = {]])
 
 for i, event in ipairs(events) do
-  if i > 1 then
-    comma = ',\n'
-  else
-    comma = '\n'
+  enum_tgt:write(('\n  EVENT_%s = %u,'):format(event:upper(), i - 1))
+  names_tgt:write(('\n  {%u, "%s", EVENT_%s},'):format(#event, event, event:upper()))
+  if i == #events then  -- Last item.
+    enum_tgt:write(('\n  NUM_EVENTS = %u,'):format(i))
```

Comment:
This value should not be part of the same enum as the list of events. It is semantically incorrect and if enum is ever used for `switch()` it would force either useless `case NUM_EVENTS: assert(false);` or harmful `default:`. If you need `NUM_EVENTS` for unit tests it should be a separate anonymous `enum { NUM_EVENTS = N };`.

Reply:
@ZyX-I Thanks for checking. This was to eliminate PVS warnings:

    fileio.c:7293   Medium  V547    Expression is always true.
    fileio.c:7351   Medium  V547    Expression 'event == 100' is always false.

`event_name2nr` returns `event_T`, so (I guess) PVS infers that nothing outside of
that range could possibly be returned.

Should we just ignore the warnings using `-V547` ? Or you have another idea?

Reply:
`event_name2nr` should not return `NUM_EVENTS`. What happens there semantically is why a bunch of enums have things like VAR_UNKNOWN (i.e. values which state that item is uninitialized or invalid) and these all are *zero* for the sake of initialization sequences like using `memset(, 0,)` yielding something relatively safe. So it makes sense to have NUM_EVENTS still outside of enum, but make `event_name2nr` return new `EVENT_INVALID` instead which ideally should be zero (though given that `event_T` is used as an index it is easier to place at the end of enum).

Though I am wondering why compiler allows NUM_EVENTS there and does not even give a warning.

About UIExtension: it is still not correct semantically to use for invalid values, but it is a separate case: UIExtension is *written* enum, not generated one. You do not have length available through some compiler builtin like `enumitems(EnumType)` and there is no list to query length like `#events`. I remember moving length out of enum somewhere, but it was ugly `#define …LEN ((size_t)LAST_ITEM+1)` with a warning to change the define at the end of the corresponding enum and AFAIR is either gone or in one of the non-merged branches (like lua to viml translator).

Reply:
Good points!